### PR TITLE
[Adjustments] Payment fee adjustment

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Following #6858

Updates the associations on payment fee adjustments so they are now primarily associated with the payment object instead of the order, reducing the required polymorphic associations from 3 to 2.

#### What should we test?

Payment fees are applied to orders at checkout as normal. Payment fees show up in admin area and reports as normal.

#### Release notes

Changelog Category: Technical changes

### CODE REVIEW

**Review by andrewpbrett — APPROVED** [review]
This looks pretty sensible in general; the changes to have `Order` return `all_adjustments` are so that these adjustments that were previously associated with the order, and are now associated with the payment, still get returned, right? 

Could we take it as tech debt to eventually get rid of `all_adjustments` and just have `adjustments` return ... all adjustments? :trollface: I don't know, maybe that's an important distinction to keep, but as someone coming it without context, it's not really clear what one does versus the other.

File: app/models/spree/payment.rb [hunk 1]
```diff
@@ -126,6 +126,10 @@ def payment_source
       res || payment_method
     end
 
+    def adjustment
+      @adjustment ||= adjustments.first
+    end
```
[the remark this answers is not in the document]
Comment (sauloperez) [reply app/models/spree/payment.rb:131]: or even ensure there's only one at DB-level.

File: app/services/order_checkout_restart.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ def call
     clear_shipments
     clear_payments
 
-    order.reload
+    order.reload.update!
```
Comment (sauloperez) [inline app/services/order_checkout_restart.rb:16]: why is this needed?

Comment (Matt-Yorkley) [reply app/services/order_checkout_restart.rb:16]: This is related to callbacks hell with adjustments. Previously, the payment fee adjustment was being updated in this step, and a callback on adjustments (which only affects adjustments directly on the order) was calling `order.update!`.

Moving the payment fee adjustment to the payment means that callback is now not triggered, which means `order.update!` is not called automatically. In this case we're changing the order's contents (in the lines above) and `order.update!` actually does need to be called, or the totals will not be updated.

File: app/models/spree/payment.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/models/spree/payment.rb]: So the issue is that payments actually _do_ have only one adjustment, but the interface for adding an adjustment to an object calls `adjustments.create()` (adjustments plural). I've adapted it a bit and set the Payment association to `has_one :adjustment` https://github.com/openfoodfoundation/openfoodnetwork/pull/6927/commits/9f0ca1aca53b5ccde3ee2590eb1d30bd735f3be5

Not sure if it's better or worse :see_no_evil:

**Review by sauloperez — APPROVED** [review]
That last commit seems good enough to me :+1: I prefer this explicit change to a wrong abstraction.

**Comment by Matt-Yorkley** [comment]
Rebased / resolved conflict. This one can go to test-ready right?

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,
Thanks for the notes on what to test.

Plaving orders on:
- shopfront
- backoffice

Apply the correct payment fee on the order. Checked this as well on different payment methods:
- cash -> OK
- Stripe SCA -> OK
- Paypal -> OK

Verified whether the fees are correctly displayed on:
- admin/reports -> yup, transaction fees appear under `order_management/reports/enterprise_fee_summary `

image

- admin/orders

These fees appear correctly, when editing an order (appearing on the Order Adjustments section):

image

However, they seem to be missing from the **Adjustments** tab. Before this PR this was the case:

image

But now they're gone:

image

I'm not sure this is intended?

**Comment by Matt-Yorkley** [comment]
Nice catch! They should be shown now :ok_hand:

**Comment by sauloperez** [comment]
IMO we should cover that regression with a test so we don't put the pressure on @filipefurtad0 next time, at least for this one :sweat_smile:

**Comment by Matt-Yorkley** [comment]
I added a quick test into that last commit and force-pushed :+1:

**Comment by filipefurtad0** [comment]
Aright, the enterprise fee is appearing on the adjustments table:

image

Thanks for the quick fix and the spec! :tada:
